Library/Homebrew/cask/cask_loader.rb

```diff
@@ -464,21 +439,51 @@ def from_h_gsubs(value, appdir)
 
     # Loader which tries loading casks from tap paths, failing
     # if the same token exists in multiple taps.
-    class FromAmbiguousTapPathLoader < FromPathLoader
+    class FromNameLoader < FromTapLoader
+      sig {
+        params(ref: T.any(String, Pathname, Cask, URI::Generic), warn: T::Boolean)
+          .returns(T.nilable(T.attached_class))
+      }
       def self.try_new(ref, warn: false)
-        case (possible_tap_casks = CaskLoader.tap_paths(ref)).count
+        return unless ref.is_a?(String)
+        return if ref.include?("/")
+
+        token = ref
+
+        loaders = Tap.map { |tap| super("#{tap}/#{token}", warn: warn) }
+                     .compact
+                     .select { _1.path.exist? }
+
+        case loaders.count
         when 1
-          new(possible_tap_casks.first)
+          loaders.first
         when 2..Float::INFINITY
-          loaders = possible_tap_casks.map(&FromPathLoader.method(:new))
-          raise TapCaskAmbiguityError.new(ref, loaders)
+          default_tap_loaders, other_loaders = *loaders.partition { _1.tap.core_cask_tap? }
+          default_tap_loader = default_tap_loaders.first if default_tap_loaders.count
+
+          # Put default tap last so that the error message always recommends
+          # using the fully-qualified name for non-default taps.
+          taps = other_loaders.map(&:tap) + default_tap_loaders.map(&:tap)
+
+          error = TapCaskAmbiguityError.new(token, taps)
+
+          raise error unless default_tap_loader
+
+          opoo error if warn
```

Comment:
We generally don't encourage `homebrew/(core|cask)/package`, so not sure what the value of this warning is. If we do it, it should be done consistently for both formulae and casks.

Reply:
A good example of this is `full_name` strips out `homebrew/core` prefix

Reply:
I removed the warning here when a cask exists in the default tap. This way (non-)API loading will have the same output.